# Amphi 4 - Algorithms and Complexity

# 1. Recurrence

Recurrence is calling a function inside itself for simpler arguments. We can use recurrence to shorten our code or to replace loops.

The following example shows how to compute $n!$ using loops.

In [26]:
def factorial(n):
    if n == 0:
        return 1
    result = 1
    for i in range(1, n+1):
        result = result * i
    return result

print(factorial(50))

30414093201713378043612608166064768844377641568960512000000000000


We can use recurrence instead

In [27]:
def factorial2(n):
    if n == 0:
        return 1
    return factorial(n-1)*n

print(factorial(50))

30414093201713378043612608166064768844377641568960512000000000000


We can compare running time for these two algorithms by launching each of them 1000 times.

In [29]:
import time
t1 = time.time()
for i in range(1000):
    factorial(50)
print("Using loop took " + str(time.time() - t1) + " seconds")

t2 = time.time()
for i in range(1000):
    factorial2(50)
print("Using recurrence took " + str(time.time() - t2) + " seconds")

Using loop took 0.00500011444092 seconds
Using recurrence took 0.00499987602234 seconds


The two function has almost the same running time.

## 1.1 An example: Fibonacci sequence

We take an example to illustrate an abused use of recurrence. We want to compute the $n^{th}$ Fibonacci number $F_n$, where
$$
F_1 = 1; F_2 = 1
$$
$$
F_n = F_{n-1} + F_{n-2}, n \geq 3
$$

In [24]:
def Fibonacci(n):
    if n == 1 or n == 2:
        return 1
    return Fibonacci(n - 1) + Fibonacci(n - 2)

import time
t1 = time.time()
print(Fibonacci(30))
print("Abusing recurrence took " + str(time.time() - t1) + " seconds.")

832040
Abusing recurrence took 0.173000097275 seconds.


To look at how many times the function has been called for each $n$, we can use a counter like this

In [23]:
counter = [0] * 31 #Remind: [0] * 31 = [0, 0, 0, ...., 0] (31 times)

def Fibonacci(n):
    counter[n] += 1
    if n == 1 or n == 2:
        return 1
    return Fibonacci(n - 1) + Fibonacci(n - 2)

print(Fibonacci(30))
print(counter)
print(sum(counter[3:]))

832040
[0, 317811, 514229, 317811, 196418, 121393, 75025, 46368, 28657, 17711, 10946, 6765, 4181, 2584, 1597, 987, 610, 377, 233, 144, 89, 55, 34, 21, 13, 8, 5, 3, 2, 1, 1]
832039


That means, to compute the $30^{th}$ Fibonacci number, we have to call and compute 317811 times **Fibonacci(1)**, 514229 times **Fibonacci(2)**, and do 832039 additions, which is a waste of computation.

## 1.2 Time Complexity and Space Complexity

If an addition can be done in $m$ seconds, then the necessary time for computing **Fibonacci(30)** is something propotional to 832039$m$. In general, **Fibonacci(n)** can be doned in a period proportional to $\left( \frac{1+\sqrt{5}}{2} \right)^n$ seconds. (Try proving it as an exercise)

When Python compute a function, it will reserve some space to store local variables of the function. And when the function returns, it will free all spaces for local variables. Using the **Fibonacci** function, during the call to **Fibonacci(n)** there are at most **n** spaces that are busy at the same time, corresponding to this call:

<center>**Fibonacci(n) $\to$ Fibonacci(n-1) $\to$ ... $\to$ Fibonacci(2) $\to$ Fibonacci(1)**</center>

These two quantities gives us notion about Time Complexity and Space Complexity.

### Negociation Between Time Complexity and Space Complexity

We can use spaces (memory) to negociate with time complexity. For example, the following implementation reduces the time to compute the $n^{th}$ Fibonacci number. 

In [19]:
computed_fibonaccis = [0, 1, 1]

def Fibonacci2(n):
    if n < len(computed_fibonaccis):
        return computed_fibonaccis[n]
    else:
        for j in range(len(computed_fibonaccis), n + 1):
            computed_fibonaccis.append(computed_fibonaccis[j - 1] + computed_fibonaccis[j - 2])
    return(computed_fibonaccis[n])

print(Fibonacci2(30))

832040


In [22]:
import time
t1 = time.time()
print(Fibonacci2(30))
print("Using recurrence with more memory took " + str(time.time() - t1) + " seconds.")

832040
Using recurrence with more memory took 0.0 seconds.


The algorithm **Fibonacci2** took a running time at max proportional to $N$ instead of some exponent function of $N$. Space complexity is still proportional to $N$.

We can use another intelligent way by looping on 2 variables.

In [31]:
def Fibonacci3(n):
    if n == 1 or n == 2:
        return 1
    a = 1
    b = 1
    for i in range(3, n + 1):
        c = a + b
        a = b
        b = c
    return b

print(Fibonacci3(30))

832040


The algorith **Fibonacci3** took a running time proportional to $N$ and a space complexity constant (only 3 local variables are needed everytime).

***Question: When should we use Fibonacci1, Fibonacci2, Fibonacci3?***

## 1.3 Bit Complexity

## 1.4 Notations

# 2. Bit Complexity of Basic Arithmetic Operations

## 2.1 Integer Addition and Substraction

## 2.2 Integer Multiplication

## 2.3 Integer Division

## 2.4 Integer Exponentation

## 2.5 Real Numbers

# 3. Complexity of Operations on Matrices

## 3.1 Matrix Multiplication

## 3.2 Matrix Inverse

# 4. Divide and Conquest

## 4.1 Divide and Conquest. Master Theorem

## 4.2 Application to Integer Multiplication and Matrix Multiplication

# 5. Basic Algorithms and Their Complexity

## 5.1 Search

## 5.2 Sorting

## 5.3 Graph Exploring

# 6. Parallel Programming with Pool in Python